File: infrastructure-python/stacks/ecs_app_stack.py
"""ECS Fargate application stack for SpatialPower (Python CDK).

A 1:1 port of infrastructure/lib/ecs-app-stack.ts. It imports the shared
VPC / subnets / security groups / cluster / ALB listener / IAM role, provisions
two FIFO SQS queues (work + error) for the async sparrpowR jobs, and runs the
app as a Fargate service behind the shared ALB. The real task definition
(frontend + backend + queue + firelens) is rendered and registered by the
deploy-app GitHub workflow from .github/aws/web.yml; the container defined here
is only a placeholder so the service can be created/updated by CDK.
"""
from aws_cdk import (
    Stack,
    Tags,
    Duration,
    RemovalPolicy,
    TimeZone,
    CfnOutput,
    Arn,
    ArnFormat,
    Environment,
    aws_ec2 as ec2,
    aws_ecs as ecs,
    aws_iam as iam,
    aws_logs as logs,
    aws_sqs as sqs,
    aws_elasticloadbalancingv2 as elbv2,
    aws_ssm as ssm,
    aws_applicationautoscaling as appscaling,
)
from constructs import Construct


class EcsAppStack(Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        env: Environment,
        stack_name: str,
        description: str,
        tier: str,
        app_name: str,
        app_namespace: str,
        app_service: str,
        app_domain: str,
        app_path_prefix: str,
        vpc_id: str,
        subnet_ids: list[str],
        security_group_ids: list[str],
        cluster_arn: str,
        listener_arn: str,
        app_role_arn: str,
        listener_rule_priority: int,
        health_check_path: str,
        grace_period: int,
        cpu: int,
        memory: int,
        desired_count: int,
        container_port: int,
        non_prod_schedule: bool,
        scheduled_min_capacity: int,
        scheduled_max_capacity: int,
    ) -> None:
        super().__init__(
            scope,
            construct_id,
            env=env,
            stack_name=stack_name,
            description=description,
        )

        # Stack-level tags
        Tags.of(self).add("EnvironmentTier", tier)
        Tags.of(self).add("ResourceName", f"{tier}-{app_name}")
        Tags.of(self).add("ManagedBy", "cdk")
        Tags.of(self).add("CreatedBy", "cdk")
        Tags.of(self).add("Project", "dceg-analysistools")
        Tags.of(self).add("ApplicationName", app_name)

        # Import existing shared resources
        vpc = ec2.Vpc.from_lookup(self, "Vpc", vpc_id=vpc_id)

        subnets = [
            ec2.Subnet.from_subnet_id(self, f"Subnet{i}", sid)
            for i, sid in enumerate(subnet_ids)
        ]

        security_groups = [
            ec2.SecurityGroup.from_security_group_id(self, f"SG{i}", sg_id)
            for i, sg_id in enumerate(security_group_ids)
        ]

        cluster_name = Arn.split(
            cluster_arn, ArnFormat.SLASH_RESOURCE_NAME
        ).resource_name
        assert cluster_name is not None
        cluster = ecs.Cluster.from_cluster_attributes(
            self,
            "Cluster",
            cluster_name=cluster_name,
            cluster_arn=cluster_arn,
            vpc=vpc,
            security_groups=security_groups,
        )

        execution_role = iam.Role.from_role_arn(self, "ExecutionRole", app_role_arn)
        task_role = iam.Role.from_role_arn(self, "TaskRole", app_role_arn)

        listener = elbv2.ApplicationListener.from_application_listener_attributes(
            self,
            "Listener",
            listener_arn=listener_arn,
            security_group=security_groups[0],
        )

        # ---------------------------------------------------------------------
        # Async job queues (SQS).
        #
        # SpatialPower submits long-running sparrpowR jobs to a work queue; the
        # queue-worker container polls it, runs the R computation, and writes
        # results to S3. Failed jobs are forwarded to a separate error queue.
        # Both are FIFO queues because the app supplies MessageGroupId /
        # MessageDeduplicationId on every send (see server/app.js,
        # queue-worker.js).
        # ---------------------------------------------------------------------
        error_queue = sqs.Queue(
            self,
            "ErrorQueue",
            queue_name=f"{tier}-{app_name}-error.fifo",
            fifo=True,
            content_based_deduplication=False,
            retention_period=Duration.days(14),
            visibility_timeout=Duration.seconds(900),
        )

        work_queue = sqs.Queue(
            self,
            "WorkQueue",
            queue_name=f"{tier}-{app_name}.fifo",
            fifo=True,
            content_based_deduplication=False,
            retention_period=Duration.days(4),
            visibility_timeout=Duration.seconds(900),
        )

        # Grant the shared task role access to the queues. APP_ROLE_ARN is the
        # shared analysistools task role, imported above with from_role_arn
        # (mutable by default), so CDK attaches scoped inline policies for THESE
        # queues only -- the same mechanism that already adds the per-app
        # CloudWatch log-group grants to that role. The backend sends to the
        # work queue; the worker consumes the work queue and forwards failures
        # to the error queue.
        work_queue.grant_send_messages(task_role)
        work_queue.grant_consume_messages(task_role)
        error_queue.grant_send_messages(task_role)

        # CloudWatch log group
        log_group = logs.LogGroup(
            self,
            "WebLogGroup",
            log_group_name=f"/{app_namespace}/{tier}/{app_name}/web",
            retention=logs.RetentionDays.SIX_MONTHS,
            removal_policy=RemovalPolicy.DESTROY,
        )

        # ---------------------------------------------------------------------
        # Placeholder task definition.
        #
        # As in the analysistools-portal reference stack, the real task definition
        # (frontend + backend + queue + firelens) is rendered and registered by
        # the deploy-app GitHub workflow from .github/aws/web.yml. This
        # placeholder only exists so the service can be created/updated by CDK;
        # the CfnService override below pins the service to the task-definition
        # *family* so CDK never reverts the workflow-registered revision.
        # ---------------------------------------------------------------------
        task_def = ecs.FargateTaskDefinition(
            self,
            "WebTaskDef",
            family=f"{tier}-{app_name}-{app_service}",
            cpu=cpu,
            memory_limit_mib=memory,
            execution_role=execution_role,
            task_role=task_role,
        )

        task_def.add_container(
            "WebContainer",
            container_name="frontend",
            image=ecs.ContainerImage.from_registry("nginx:alpine"),
            essential=True,
            port_mappings=[
                ecs.PortMapping(
                    container_port=container_port,
                    host_port=container_port,
                    protocol=ecs.Protocol.TCP,
                )
            ],
            logging=ecs.LogDrivers.aws_logs(
                log_group=log_group,
                stream_prefix="frontend",
            ),
        )

        # Target group
        tg = elbv2.ApplicationTargetGroup(
            self,
            "WebTG",
            target_group_name=f"{tier}-{app_name}-{app_service}",
            port=container_port,
            protocol=elbv2.ApplicationProtocol.HTTP,
            target_type=elbv2.TargetType.IP,
            vpc=vpc,
            health_check=elbv2.HealthCheck(
                enabled=True,
                path=health_check_path,
                port=str(container_port),
                healthy_http_codes="200-499",
            ),
        )

        # ALB listener rule: route the app's host + path prefix to this service.
        listener.add_target_groups(
            "WebListenerRule",
            target_groups=[tg],
            conditions=[
                elbv2.ListenerCondition.host_headers([app_domain]),
                elbv2.ListenerCondition.path_patterns(
                    [app_path_prefix, f"{app_path_prefix}/*"]
                ),
            ],
            priority=listener_rule_priority,
        )

        # Fargate service
        service = ecs.FargateService(
            self,
            "WebService",
            service_name=f"{tier}-{app_name}-{app_service}",
            cluster=cluster,
            task_definition=task_def,
            desired_count=desired_count,
            security_groups=security_groups,
            vpc_subnets=ec2.SubnetSelection(subnets=subnets),
            assign_public_ip=False,
            enable_ecs_managed_tags=True,
            enable_execute_command=True,
            circuit_breaker=ecs.DeploymentCircuitBreaker(rollback=True),
            health_check_grace_period=Duration.seconds(grace_period),
            propagate_tags=ecs.PropagatedTagSource.TASK_DEFINITION,
        )

        service.attach_to_application_target_group(tg)

        # Prevent CDK from reverting task definitions registered by deploy-app workflow
        cfn_service = service.node.default_child
        cfn_service.add_property_override(
            "TaskDefinition", f"{tier}-{app_name}-{app_service}"
        )
        cfn_service.add_property_deletion_override("DesiredCount")

        # Scheduled auto-scaling (non-prod: scale to 0 nights/weekends)
        if non_prod_schedule:
            scalable = service.auto_scale_task_count(
                min_capacity=0,
                max_capacity=scheduled_max_capacity,
            )

            scalable.scale_on_schedule(
                "ScaleOut",
                schedule=appscaling.Schedule.cron(
                    hour="7", minute="0", week_day="MON-FRI"
                ),
                min_capacity=scheduled_min_capacity,
                max_capacity=scheduled_max_capacity,
                time_zone=TimeZone.AMERICA_NEW_YORK,
            )

            scalable.scale_on_schedule(
                "ScaleIn",
                schedule=appscaling.Schedule.cron(
                    hour="19", minute="0", week_day="MON-FRI"
                ),
                min_capacity=0,
                max_capacity=0,
                time_zone=TimeZone.AMERICA_NEW_YORK,
            )

        # SSM parameters for deploy-app workflow
        ssm.StringParameter(
            self,
            "SsmEcsCluster",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/ecs_cluster",
            string_value=cluster_name,
        )

        ssm.StringParameter(
            self,
            "SsmEcsWebTask",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/ecs_web_task",
            string_value=f"{tier}-{app_name}-{app_service}",
        )

        ssm.StringParameter(
            self,
            "SsmEcsWebService",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/ecs_web_service",
            string_value=f"{tier}-{app_name}-{app_service}",
        )

        ssm.StringParameter(
            self,
            "SsmRoleArn",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/role_arn",
            string_value=app_role_arn,
        )

        ssm.StringParameter(
            self,
            "SsmQueueName",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/queue_name",
            string_value=work_queue.queue_name,
        )

        ssm.StringParameter(
            self,
            "SsmQueueUrl",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/queue_url",
            string_value=work_queue.queue_url,
        )

        ssm.StringParameter(
            self,
            "SsmQueueErrorUrl",
            parameter_name=f"/{app_namespace}/{tier}/{app_name}/queue_error_url",
            string_value=error_queue.queue_url,
        )

        # Stack outputs
        CfnOutput(
            self,
            "WebServiceName",
            value=service.service_name,
            description="ECS Service Name",
        )

        CfnOutput(
            self,
            "WebTaskDefArn",
            value=task_def.task_definition_arn,
            description="Task Definition ARN",
        )

        CfnOutput(
            self,
            "TargetGroupArn",
            value=tg.target_group_arn,
            description="Target Group ARN",
        )

        CfnOutput(
            self,
            "WorkQueueUrl",
            value=work_queue.queue_url,
            description="SQS work queue URL",
        )

        CfnOutput(
            self,
            "ErrorQueueUrl",
            value=error_queue.queue_url,
            description="SQS error queue URL",
        )

File: infrastructure/lib/ecs-app-stack.ts
import * as cdk from "aws-cdk-lib";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import * as sqs from "aws-cdk-lib/aws-sqs";
import * as elbv2 from "aws-cdk-lib/aws-elasticloadbalancingv2";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as appscaling from "aws-cdk-lib/aws-applicationautoscaling";
import { Construct } from "constructs";

export interface EcsAppStackProps extends cdk.StackProps {
  tier: string;
  appName: string;
  appNamespace: string;
  appService: string;
  appDomain: string;
  appPathPrefix: string;

  vpcId: string;
  subnetIds: string[];
  securityGroupIds: string[];
  clusterArn: string;
  listenerArn: string;
  appRoleArn: string;

  listenerRulePriority: number;
  healthCheckPath: string;
  gracePeriod: number;

  cpu: number;
  memory: number;
  desiredCount: number;
  containerPort: number;

  nonProdSchedule: boolean;
  scheduledMinCapacity: number;
  scheduledMaxCapacity: number;
}

export class EcsAppStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: EcsAppStackProps) {
    super(scope, id, props);

    const {
      tier,
      appName,
      appNamespace,
      appService,
      appDomain,
      appPathPrefix,
      vpcId,
      subnetIds,
      securityGroupIds,
      clusterArn,
      listenerArn,
      appRoleArn,
      listenerRulePriority,
      healthCheckPath,
      gracePeriod,
    } = props;

    // Stack-level tags
    cdk.Tags.of(this).add("EnvironmentTier", tier);
    cdk.Tags.of(this).add("ResourceName", `${tier}-${appName}`);
    cdk.Tags.of(this).add("ManagedBy", "cdk");
    cdk.Tags.of(this).add("CreatedBy", "cdk");
    cdk.Tags.of(this).add("Project", "dceg-analysistools");
    cdk.Tags.of(this).add("ApplicationName", appName);

    // Import existing shared resources
    const vpc = ec2.Vpc.fromLookup(this, "Vpc", { vpcId });

    const subnets = subnetIds.map((sid, i) =>
      ec2.Subnet.fromSubnetId(this, `Subnet${i}`, sid)
    );

    const securityGroups = securityGroupIds.map((sgId, i) =>
      ec2.SecurityGroup.fromSecurityGroupId(this, `SG${i}`, sgId)
    );

    const clusterName = cdk.Arn.split(
      clusterArn,
      cdk.ArnFormat.SLASH_RESOURCE_NAME
    ).resourceName!;
    const cluster = ecs.Cluster.fromClusterAttributes(this, "Cluster", {
      clusterName,
      clusterArn,
      vpc,
      securityGroups,
    });

    const executionRole = iam.Role.fromRoleArn(this, "ExecutionRole", appRoleArn);
    const taskRole = iam.Role.fromRoleArn(this, "TaskRole", appRoleArn);

    const listener = elbv2.ApplicationListener.fromApplicationListenerAttributes(
      this,
      "Listener",
      {
        listenerArn,
        securityGroup: securityGroups[0],
      }
    );

    // Async job queues (FIFO)
    const errorQueue = new sqs.Queue(this, "ErrorQueue", {
      queueName: `${tier}-${appName}-error.fifo`,
      fifo: true,
      contentBasedDeduplication: false,
      retentionPeriod: cdk.Duration.days(14),
      visibilityTimeout: cdk.Duration.seconds(900),
    });

    const workQueue = new sqs.Queue(this, "WorkQueue", {
      queueName: `${tier}-${appName}.fifo`,
      fifo: true,
      contentBasedDeduplication: false,
      retentionPeriod: cdk.Duration.days(4),
      visibilityTimeout: cdk.Duration.seconds(900),
    });

    workQueue.grantSendMessages(taskRole);
    workQueue.grantConsumeMessages(taskRole);
    errorQueue.grantSendMessages(taskRole);

    // CloudWatch log group
    const logGroup = new logs.LogGroup(this, "WebLogGroup", {
      logGroupName: `/${appNamespace}/${tier}/${appName}/web`,
      retention: logs.RetentionDays.SIX_MONTHS,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Placeholder task definition; real one registered by deploy-app workflow
    const taskDef = new ecs.FargateTaskDefinition(this, "WebTaskDef", {
      family: `${tier}-${appName}-${appService}`,
      cpu: props.cpu,
      memoryLimitMiB: props.memory,
      executionRole,
      taskRole,
    });

    taskDef.addContainer("WebContainer", {
      containerName: "frontend",
      image: ecs.ContainerImage.fromRegistry("nginx:alpine"),
      essential: true,
      portMappings: [
        {
          containerPort: props.containerPort,
          hostPort: props.containerPort,
          protocol: ecs.Protocol.TCP,
        },
      ],
      logging: ecs.LogDrivers.awsLogs({
        logGroup,
        streamPrefix: "frontend",
      }),
    });

    // Target group
    const tg = new elbv2.ApplicationTargetGroup(this, "WebTG", {
      targetGroupName: `${tier}-${appName}-${appService}`,
      port: props.containerPort,
      protocol: elbv2.ApplicationProtocol.HTTP,
      targetType: elbv2.TargetType.IP,
      vpc,
      healthCheck: {
        enabled: true,
        path: healthCheckPath,
        port: String(props.containerPort),
        // healthyHttpCodes omitted — defaults to "200" so a broken SPA returning
        // 4xx is treated as unhealthy and pulled from rotation.
      },
    });

    // ALB listener rule: route the app's host + path prefix to this service.
    listener.addTargetGroups("WebListenerRule", {
      targetGroups: [tg],
      conditions: [
        elbv2.ListenerCondition.hostHeaders([appDomain]),
        elbv2.ListenerCondition.pathPatterns([
          appPathPrefix,
          `${appPathPrefix}/*`,
        ]),
      ],
      priority: listenerRulePriority,
    });

    // Fargate service
    const service = new ecs.FargateService(this, "WebService", {
      serviceName: `${tier}-${appName}-${appService}`,
      cluster,
      taskDefinition: taskDef,
      // Create the placeholder service with 0 tasks. The nginx:alpine placeholder
      // returns 404 at the health-check path and the matcher is strict 200, so a
      // running placeholder would trip the deployment circuit breaker on first
      // CREATE. deploy-app scales the service to the real count (its
      // `update-service --desired-count 1`) with the real image, which is healthy.
      // NOTE: infra updates reset the count to 0 — always run deploy-app after
      // deploy-infrastructure (the documented infra -> app order).
      desiredCount: 0,
      securityGroups,
      vpcSubnets: { subnets },
      assignPublicIp: false,
      enableECSManagedTags: true,
      enableExecuteCommand: true,
      circuitBreaker: { rollback: true },
      healthCheckGracePeriod: cdk.Duration.seconds(gracePeriod),
      propagateTags: ecs.PropagatedTagSource.TASK_DEFINITION,
    });

    service.attachToApplicationTargetGroup(tg);

    // Pin the service to the task-definition family so CDK doesn't revert the
    // revisions deploy-app registers. (DesiredCount stays at the template's 0
    // here; deploy-app sets the running count via update-service.)
    const cfnService = service.node.defaultChild as ecs.CfnService;
    cfnService.addPropertyOverride(
      "TaskDefinition",
      `${tier}-${appName}-${appService}`
    );

    // Scheduled auto-scaling (non-prod: scale to 0 nights/weekends)
    if (props.nonProdSchedule) {
      const scalable = service.autoScaleTaskCount({
        minCapacity: 0,
        maxCapacity: props.scheduledMaxCapacity,
      });

      scalable.scaleOnSchedule("ScaleOut", {
        schedule: appscaling.Schedule.cron({
          hour: "7",
          minute: "0",
          weekDay: "MON-FRI",
        }),
        minCapacity: props.scheduledMinCapacity,
        maxCapacity: props.scheduledMaxCapacity,
        timeZone: cdk.TimeZone.AMERICA_NEW_YORK,
      });

      scalable.scaleOnSchedule("ScaleIn", {
        schedule: appscaling.Schedule.cron({
          hour: "19",
          minute: "0",
          weekDay: "MON-FRI",
        }),
        minCapacity: 0,
        maxCapacity: 0,
        timeZone: cdk.TimeZone.AMERICA_NEW_YORK,
      });
    }

    // SSM parameters for deploy-app workflow
    new ssm.StringParameter(this, "SsmEcsCluster", {
      parameterName: `/${appNamespace}/${tier}/${appName}/ecs_cluster`,
      stringValue: clusterName,
    });

    new ssm.StringParameter(this, "SsmEcsWebTask", {
      parameterName: `/${appNamespace}/${tier}/${appName}/ecs_web_task`,
      stringValue: `${tier}-${appName}-${appService}`,
    });

    new ssm.StringParameter(this, "SsmEcsWebService", {
      parameterName: `/${appNamespace}/${tier}/${appName}/ecs_web_service`,
      stringValue: `${tier}-${appName}-${appService}`,
    });

    new ssm.StringParameter(this, "SsmRoleArn", {
      parameterName: `/${appNamespace}/${tier}/${appName}/role_arn`,
      stringValue: appRoleArn,
    });

    new ssm.StringParameter(this, "SsmQueueName", {
      parameterName: `/${appNamespace}/${tier}/${appName}/queue_name`,
      stringValue: workQueue.queueName,
    });

    new ssm.StringParameter(this, "SsmQueueErrorUrl", {
      parameterName: `/${appNamespace}/${tier}/${appName}/queue_error_url`,
      stringValue: errorQueue.queueUrl,
    });

    // Stack outputs
    new cdk.CfnOutput(this, "WebServiceName", {
      value: service.serviceName,
      description: "ECS Service Name",
    });

    new cdk.CfnOutput(this, "WebTaskDefArn", {
      value: taskDef.taskDefinitionArn,
      description: "Task Definition ARN",
    });

    new cdk.CfnOutput(this, "TargetGroupArn", {
      value: tg.targetGroupArn,
      description: "Target Group ARN",
    });

    new cdk.CfnOutput(this, "WorkQueueUrl", {
      value: workQueue.queueUrl,
      description: "SQS work queue URL",
    });

    new cdk.CfnOutput(this, "ErrorQueueUrl", {
      value: errorQueue.queueUrl,
      description: "SQS error queue URL",
    });
  }
}
